contract/contracts/Investor/Investor.sol

```diff
@@ -0,0 +1,129 @@
+
+contract Investor is ERC20Upgradeable, OwnableUpgradeable, PausableUpgradeable, ReentrancyGuardUpgradeable {
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    EnumerableSet.AddressSet private investors;
+    mapping(address => MintAgreement) private mintAgreements;
+
+    event Minted(address indexed investor, uint256 amount);
+    event MintAgreementAdded(address indexed investor, uint256 amount, bool isActive);
+    event MintAgreementUpdated(address indexed investor, uint256 amount, bool isActive);
+    event MintAgreementRemoved(address indexed investor);
+    event DividendDistributed(address indexed investor, uint256 amount);
+    event MintingStateChanged(bool state);
+
+    function initialize(string calldata _name, string calldata _symbol, MintAgreement[] calldata _mintAgreements)
+        external
+        initializer
+    {
+        __ERC20_init(_name, _symbol);
+        __Ownable_init(msg.sender);
+        __ReentrancyGuard_init();
+        __Pausable_init();
+        for (uint256 i = 0; i < _mintAgreements.length; i++) {
+            require(_mintAgreements[i].investor != address(0), "Invalid investor address");
+
+            MintAgreement memory agreement = _mintAgreements[i];
+            investors.add(agreement.investor);
+            mintAgreements[agreement.investor] = agreement;
+        }
+    }
+
+    function mint() external onlyOwner whenNotPaused nonReentrant {
+        for (uint256 i = 0; i < investors.length(); i++) {
+            address investor = investors.at(i);
+            MintAgreement storage agreement = mintAgreements[investor];
+            if (agreement.isActive) {
+                agreement.totalMinted += agreement.amount;
+                agreement.lastMinted = block.timestamp;
+                _mint(investor, agreement.amount);
+
+                emit Minted(investor, agreement.amount);
+            }
+        }
+    }
+
+    function distributeDividends() external payable onlyOwner nonReentrant {
+        uint256 totalSupply = totalSupply();
+        require(totalSupply > 0, "No tokens minted");
+
+        for (uint256 i = 0; i < investors.length(); i++) {
+            address account = investors.at(i);
+            uint256 balance = balanceOf(account);
+            uint256 dividend = (msg.value * balance) / totalSupply;
+            payable(account).transfer(dividend);
+
+            emit DividendDistributed(account, dividend);
+        }
+    }
+
+    function addMintAgreement(address _investor, uint256 _amount, bool _isActive) external onlyOwner whenNotPaused {
+        require(_investor != address(0), "Invalid investor address");
+        require(!investors.contains(_investor), "Investor already exists");
+        require(_amount > 0, "Invalid amount");
+
+        MintAgreement storage agreement = mintAgreements[_investor];
+        agreement.investor = _investor;
+        agreement.amount = _amount;
+        agreement.isActive = _isActive;
+
```

Comment:
My only request is here. I would like to be able to initiate a user with an amount of SHERS. Something like what you might consider a "signing bonus".